### PR TITLE
Alerting: Improve threshold displays

### PR DESCRIPTION
**Pull request by gillesdemey**
**What is this feature?**

Introduces a DAG to extract correct thresholds for the alert create / edit display. 

Imagine two data queries linking to an expression, which in turn is connected to two other expressions. This is a DAG data structure.

We then have to walk the graph backwards to find our source nodes and find the data queries where we have to show the configured thresholds.

```
    ┌───┐         ┌───┐
    │ A ├┐      ┌─▶ D │
    └───┘│ ┌───┐│ └───┘
         ├─▶ C ├┤
    ┌───┐│ └───┘│ ┌───┐
    │ B ├┘      └─▶ E │
    └───┘         └───┘
```

- [x] Does not support threshold ranges yet but I intend to add that to this PR.
- [x] Doesn't work with multiple thresholds in a single expression (classic condition).
- [x] ~Drag and drop behavior of the threshold indicator is still not working.~

On the last todo item I'm thinking of simply removing its functionality – it's a tough one to implement properly with the addition of the `threshold` expression (its data structure is quite different from `classic_condition`). 

Any thoughts?

EDIT: discussed with the team and we agree that adding in the drag handle behavior is complicated and adds very little benefit over the text input.

**Notes for reviewers**

Setting this to draft so I can resolve the last of the TODOs, add some code comments to explain the intent of the DAG and fix some tests.

image

### CODE REVIEW

File: public/app/features/alerting/unified/components/rule-editor/util.ts [hunk 1]
```diff
@@ -108,3 +114,162 @@ export function warningFromSeries(series: DataFrame[]): Error | undefined {
...
+  {
+    config: ThresholdsConfig;
+    mode: GraphTresholdsStyleMode;
+  }
+>;
+
+/**
+ * This function will retrieve threshold definitions for the given array of data and expression queries.
+ *
+ * The result of this function will include
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rule-editor/util.ts:129]: This sentence seems unfinished, lol

**Review by soniaAguilarPeiron — APPROVED** [review]
Great job! 🎉

**Comment by armandgrillet** [comment]
Decision regarding backporting this: it makes sense to backport this to 9.2 as a customer reported this issue in Grafana 9.2.

**Comment by gillesdemey** [comment]
Manual back-port PRs:

`v9.3.x` https://github.com/grafana/grafana/pull/60704
`v9.2.x` <in progress>

**Pull request merged**